src/index/base.cpp

```diff
@@ -269,6 +295,38 @@ void BaseIndex::BlockConnected(const std::shared_ptr<const CBlock>& block, const
         // m_synced. Consider the case where there is a reorg and the blocks on the stale branch are
         // in the ValidationInterface queue backlog even after the sync thread has caught up to the
         // new chain tip. In this unlikely event, log a warning and let the queue clear.
+        //
+        // To allow handling reorgs, this only checks that the new block
+        // connects to ancestor of the current best block, instead of checking
+        // that it connects to directly to the current block. If there is a
+        // reorg, Rewind call below will remove existing blocks from the index
+        // before adding the new one.
+        //
+        // The other case where the new block will connect to an ancestor of the
+        // current block rather than the current block is when the m_synced flag
+        // is set to true too early. For example if the index is synced to
+        // height 100, and -reindex-chainstate option is used, there may be
+        // BlockConnected notifications for blocks 97, 98, 99, and 100 sitting
+        // in the notifications queue when m_synced gets set to true. When this
```

Comment:
In c0d1110d:

Guess that you wrote this because you saw it on an unit test?

I'm not seeing how this could be possible during the regular node initialization process. We aren't signaling any block prior to initializing the indexes.

Reply:
Ok, I think that know how that could happen. It is not related to the `Init()` call.

The background sync process reads blocks directly from the chain index (calling `NextSyncBlock()`). So, what if a block gets connected while `ThreadSync()` is being executed but the event is not dispatched on time.. (before the background sync process finishes)

1) The block connection event is added to the validation queue.
2) The index background sync process reads the block from the chain and processes it.
3) The validation queue dispatches the block connection event added in (1) to the index. Then... boom: the index received an already processed block.

Reply:
That's accurate, but I don't think ThreadSync needs to even do anything for this to happen, and no new blocks need to be connected after ThreadSync starts.

If `-reindex-chainstate` indexing completes, and ThreadSync immediately starts and exits because it detects that the index best block is equal to the chain tip, and no new blocks are connected, there could still be stale block connected notifications in the queue which will cause the index to rewind and add the blocks again. The commit 
"Avoid race, make -reindex-chainstate more efficient" e5f2641137413da5970007a9b932b5993067b550 changes index startup behavior to just ignore the stale block connected notifications while `m_ready` is `false`, instead of processing them and pointlessly rewinding and reconnecting blocks.

Reply:
Yeah ok, that is also a possibility. Great.

Could the fix be decoupled into a standalone PR? Sounds like something that we could rapidly merge independently of this PR. It fixes a bug and also makes this work smaller.

Reply:
> Yes, I am definitely looking for suggestions on how many of the first commits it would make sense to split off into a separate PR.

Ok cool. I suggested to decouple this commit only because I'm going commit by commit and it seems like an easy step forward, but will keep moving forward.

> But just to be clear, there isn't a bug here, just suboptimal behavior. If -reindex-chainstate is used and there are still block-connected notifications waiting in the queue when the index is started, it doesn't cause any index corruption. It just causes the last few blocks in the index to be rewound, and then appended again, as if there were a reorg

Yeah, we are sync. I said "bug" because of the unexpected behavior. I don't think that it was programmed to do that.